### PR TITLE
Preserve referential equality from previous result in the store

### PR DESCRIPTION
**Pull request by calebmer**
This PR implements the preservation of referential equality across reads from the store for data that has not changed. So say you have an array of `[a, b, c]` and a mutation adds a new item, `d`, to make `[a, b, c, d]`. Then the following should hold `last[0] === next[0] && last[1] === next[1] && last[2] === next[2]` while the array itself is not necessarily referentially equal (`last !== next`). This makes it easier to create optimizations in the UI layer to only re-render data which changed. Specifically this would make the implementation of `shouldComponentUpdate` in React apps much cleaner.

This PR is not yet finished, but I’d like an initial review of what’s there. I’m going to work on some other issues for a bit then come back to this. The implementation is a little messy as the `graphql-anywhere` resolver and `resultMapper` need to work together to both identify which values have previous results and if those previous results equal the new result. The implementation gets a little cleaner if we remove support for arrays whose order changes (e.g. `[a, b, c] -> [b, c, a]`), but that means we wouldn’t support the case where an item is added to the beginning of an array which is pretty important (e.g. `[a, b, c] -> [d, a, b, c]`).

What’s left to do is wiring this up with wherever it is called so that `previousResult` values actually get passed down into `diffQueryAgainstStore`.

Closes https://github.com/apollostack/apollo-client/issues/695

- [x] If this PR is a new feature, reference an issue where a consensus about the design was reached (not necessary for small changes)
- [x] Make sure all of the significant new logic is covered by tests
- [x] Rebase your changes on master so that they can be merged easily
- [x] Make sure all tests and linter rules pass
- [x] Update CHANGELOG.md with your change

### CODE REVIEW

File: src/data/readFromStore.ts [hunk 1]
```diff
@@ -170,27 +187,139 @@ Perhaps you want to use the \`returnPartialData\` option?`);
     return fieldValue;
   }
 
+  // if this is an object scalar, it must be a json blob and we have to unescape it
   if (isJsonValue(fieldValue)) {
-    // if this is an object scalar, it must be a json blob and we have to unescape it
+    // If the JSON blob is the same now as in the previous result, return the previous result to
+    // maintain referential equality.
+    if (idValue.previousResult && isEqual(idValue.previousResult[fieldName], fieldValue.json)) {
```
Comment (stubailo) [inline src/data/readFromStore.ts]: I think we can use === here instead of isEqual

File: src/data/readFromStore.ts [hunk 2]
```diff
@@ -170,27 +187,139 @@ Perhaps you want to use the \`returnPartialData\` option?`);
...
+function resultMapper (resultFields: any, idValue: IdValueWithPreviousResult) {
+  // If we had a previous result, we may be able to return that and preserve referential equality
+  if (idValue.previousResult) {
+    // Perform a shallow comparison of the result fields with the previous result. If all of
+    // the shallow fields are referentially equal to the fields of the previous result we can
+    // just return the previous result.
+    //
+    // While we do a shallow comparison of objects, we do a deep comparison of arrays.
+    const sameAsPreviousResult = Object.keys(resultFields).reduce(
+      (same, key) => {
```
Comment (stubailo) [inline src/data/readFromStore.ts]: Code style - I'd put these args on the previous line.

File: src/data/readFromStore.ts [hunk 3]
```diff
@@ -170,27 +187,139 @@ Perhaps you want to use the \`returnPartialData\` option?`);
...
+        if (!same) {
+          return false;
+        }
+
+        // Flatten out the field values before comparing them. Non-arrays will turn into singleton
+        // arrays and multi-dimensional arrays will be flattened out. Depth doesn’t matter in this
+        // case, we just need to check that all items are equal.
+        const next = flattenArray(resultFields[key]);
```
Comment (stubailo) [inline src/data/readFromStore.ts]: This would definitely pass through a false positive if the arrays had the same items in flattened form but had different shapes.

File: src/data/readFromStore.ts (hunk 2 above)
Comment (calebmer) [inline src/data/readFromStore.ts]: Yeah, this line went through multiple iterations.

File: src/data/readFromStore.ts (hunk 1 above)
Comment (stubailo) [inline src/data/readFromStore.ts]: OK I'm fine with this for now - let's do some benchmarks later and see if it is ever a problem!

File: src/data/readFromStore.ts [hunk 4]
```diff
@@ -170,27 +187,136 @@ Perhaps you want to use the \`returnPartialData\` option?`);
...
+    // While we do a shallow comparison of objects, we do a deep comparison of arrays.
+    const sameAsPreviousResult = Object.keys(resultFields).reduce((same, key) => {
+      if (!same) {
+        return false;
+      }
+
+      // Flatten out the field values before comparing them. Non-arrays will turn into singleton
+      // arrays and multi-dimensional arrays will be flattened out. Depth doesn’t matter in this
+      // case, we just need to check that all items are equal.
+      const next = flattenArray(resultFields[key]);
```
Comment (helfer) [inline src/data/readFromStore.ts]: What are the performance implications of this? I'm not too worried about deeply nested arrays at this point in time, but what about one-dimensional arrays? If we're copying every array every time to make a comparison that could amount to a fair bit of copying.

Although... I once said never optimize something you don't already know needs optimizing. So that might count here as well.

File: src/data/readFromStore.ts (hunk 1 above)
Comment (helfer) [inline src/data/readFromStore.ts]: I'm not sure I follow here. Surely once written to the store the object doesn't change, right? It's only when another request fetches the same object that the reference changes. If that's the case, then we could first check referential equality and only if that fails apply `isEqual`.

File: test/diffAgainstStore.ts [hunk 5]
```diff
@@ -346,4 +347,347 @@ describe('diffing queries against the store', () => {
...
+        c: { d: 2, e: [{ f: 3.1 }, { f: 3.2 }, { f: 3.3 }, { f: 3.4 }, { f: 3.5 }] },
+      };
+
+      const { result } = diffQueryAgainstStore({
+        store,
+        query,
+        previousResult,
+      });
+
+      assert.deepEqual(result, queryResult);
```
Comment (helfer) [inline test/diffAgainstStore.ts]: I guess this assertion wouldn't strictly be necessary, but you put it in to speed up debugging if things are still `deepEqual` but not strictly equal?

File: test/diffAgainstStore.ts [hunk 6]
```diff
@@ -346,4 +347,347 @@ describe('diffing queries against the store', () => {
       });
     });
   });
+
```
Comment (helfer) [inline test/diffAgainstStore.ts]: These tests look really neat! For some reason that makes me really happy.

**Comment by stubailo** [comment]
@helfer when I talked to Caleb he mentioned that `isEqual` checks referential equality first anyway for performance.

File: src/data/readFromStore.ts (hunk 1 above)
Comment (calebmer) [inline src/data/readFromStore.ts]: @helfer `isEqual` always does a [referential check first](https://github.com/apollostack/apollo-client/blob/e1180e332b661d684b228d24ab3d95252dabddb5/src/util/isEqual.ts#L6-L8). If you’d like we could do another one here to make the intention explicit.

File: src/data/readFromStore.ts (hunk 4 above)
Comment (calebmer) [inline src/data/readFromStore.ts]: My guess would be it isn’t a huge deal, but I’m going to change it anyway. @stubailo pointed out it could be a source of bugs by creating false positives.

File: test/diffAgainstStore.ts (hunk 5 above)
Comment (calebmer) [inline test/diffAgainstStore.ts]: Yeah, it’s more of a quick sanity check then anything else. For instance `result` might not deep equal `queryResult` if we wrote `result.a.b = 10` on line 432 but would still strict equal `previousResult`.

File: src/data/readFromStore.ts (hunk 3 above)
Comment (calebmer) [inline src/data/readFromStore.ts]: You’re right. I initially patched `isEqual` to do shallow object checks when a config flag was passed, but that felt wrong so I changed it to this. There is probably a better solution.

File: src/data/readFromStore.ts (hunk 1 above)
Comment (calebmer) [inline src/data/readFromStore.ts]: If you don’t care about referential equality when a new JSON object comes back from the API (which when a deep comparison is performed is actually the same object) we could just remove this entirely as the edge case may not be worth supporting at all. I only really added it to get [this](https://github.com/apollostack/apollo-client/pull/1136/files#diff-8cadc1a617c11332dce6d2e2cc3f5da7R657) test to pass, otherwise it’s a safe removal.

**Comment by calebmer** [comment]
Added some commits to address comments and improve file legibility. I’m going to work on getting the `previousResult` passed into the appropriate call sites for `diffQueryAgainstStore` now. I’ll ping the thread again once that’s done :+1:

**Comment by calebmer** [comment]
Ok, so I added the `previousResult` option [here](https://github.com/apollostack/apollo-client/blob/a1ae3b65f021f72f72b607cb27f9956887093886/src/core/QueryManager.ts#L364) which I *think* is all that really matters? Because it appears that is the code path which creates results for `watchQuery` which is what React uses. Whenever the store changes this code path generates a new result for the `watchQuery` observers.

Other locations which we could add a `previousResult` to is as follows. Let me know if a `previousResult` should be added to any of these locations:

1. https://github.com/apollostack/apollo-client/blob/d314e2f1ee3617d9a5715a677b1c17da77701288/src/core/QueryManager.ts#L477-L483
2. https://github.com/apollostack/apollo-client/blob/d314e2f1ee3617d9a5715a677b1c17da77701288/src/core/QueryManager.ts#L755
3. https://github.com/apollostack/apollo-client/blob/d314e2f1ee3617d9a5715a677b1c17da77701288/src/core/QueryManager.ts#L762
4. https://github.com/apollostack/apollo-client/blob/d314e2f1ee3617d9a5715a677b1c17da77701288/src/core/QueryManager.ts#L964-L970
5. https://github.com/apollostack/apollo-client/blob/d314e2f1ee3617d9a5715a677b1c17da77701288/src/data/resultReducers.ts#L44-L50

File: CHANGELOG.md [hunk 7]
```diff
@@ -15,6 +15,7 @@ Expect active development and potentially significant breaking changes in the `0
 - Remove lodash as a production dependency [PR #1122](https://github.com/apollostack/apollo-client/pull/1122)
 - Minor fix to write to `ROOT_SUBSCRIPTION` ID in the store for subscription results. [PR #1122](https://github.com/apollostack/apollo-client/pull/1127)
 - Remove `whatwg-fetch` polyfill dependency and instead warn when a global `fetch` implementation is not found. [PR #1134](https://github.com/apollostack/apollo-client/pull/1134)
+- New results from `watchQuery` are referentially equal (so `a === b`) to previous results if nothing changed in the store for a better UI integration experience when determining what changed. [PR #1136](https://github.com/apollostack/apollo-client/pull/1136)
```
Comment (stubailo) [inline CHANGELOG.md]: This misses the important point that this applies to sub-objects in the result, not just the whole result (which was true before!)

Comment (calebmer) [inline CHANGELOG.md]: Fixed :+1:

**Comment by stubailo** [comment]
We should write a blog post about this when it ships. I think it will do a lot to demonstrate the attention to detail of what we're doing.

File: src/data/readFromStore.ts [hunk 8]
```diff
@@ -228,3 +275,126 @@ an object reference. This should never happen during normal use unless you have
...
+    return true;
+  }
+  // If either `a` or `b` are not an array return false. `a` and `b` are known to not be equal
+  // here, we checked above.
+  if (!Array.isArray(a) || !Array.isArray(b)) {
+    return false;
+  }
+  // Otherwise let us compare all of the array items (which are potentially nested arrays!) to see
+  // if they are equal.
+  return a.reduce((same, item, i) => same && areNestedArrayItemsStrictlyEqual(item, b[i]), true);
```
Comment (helfer) [inline src/data/readFromStore.ts]: With this comparison we'd end up with `[1,2] === [1,2,3]` because we're only comparing up to the number of items in array a.

Comment (calebmer) [inline src/data/readFromStore.ts]: Yep, that’s my bad. Comparing length up front is probably good hygiene anyway.

**Comment by helfer** [comment]
Sweet, that's an awesome PR @calebmer!

I added `previousResult` to `getCurrentResult` as well because React Apollo uses that in a bunch of places, and I _think_ we need referential equality there as well. It definitely doesn't hurt.

Going to merge this now.

**Comment by danielgriese** [comment]
WUAT? This bugged me for a long time. Awesome! Thank you very much and sorry for the spam... just wanted to show my gratitude! :D
